packages/snaps-simulator/src/features/simulation/slice.ts

```diff
@@ -0,0 +1,170 @@
+import {
+  GenericPermissionController,
+  SubjectMetadataController,
+} from '@metamask/permission-controller';
+import { DialogType } from '@metamask/rpc-methods';
+import { IframeExecutionService } from '@metamask/snaps-controllers';
+import { Component } from '@metamask/snaps-ui';
+import {
+  SnapManifest,
+  SnapRpcHookArgs,
+  VirtualFile,
+} from '@metamask/snaps-utils';
+import {
+  createAction,
+  createSelector,
+  createSlice,
+  PayloadAction,
+} from '@reduxjs/toolkit';
+
+export enum SnapStatus {
+  Ok,
+  Loading,
+  Error,
+}
+
+export type HandlerUserInterface = {
+  type: DialogType;
+  snapId: string;
+  snapName: string;
+  node: Component;
+};
+
+type SimulationState = {
+  status: SnapStatus;
+  executionService: IframeExecutionService | null;
+  permissionController: GenericPermissionController | null;
+  subjectMetadataController: SubjectMetadataController | null;
+  manifest: VirtualFile<SnapManifest> | null;
+  sourceCode: VirtualFile<string> | null;
+  icon?: VirtualFile<string>;
+  ui?: HandlerUserInterface | null;
+  snapState: string | null;
+};
+
+export const INITIAL_STATE: SimulationState = {
+  status: SnapStatus.Loading,
+  executionService: null,
+  permissionController: null,
+  subjectMetadataController: null,
+  manifest: null,
+  sourceCode: null,
+  snapState: null,
+};
+
+const slice = createSlice({
+  name: 'simulation',
+  initialState: INITIAL_STATE,
+  reducers: {
+    setStatus(state, action: PayloadAction<SnapStatus>) {
+      state.status = action.payload;
+    },
+    setExecutionService(state, action: PayloadAction<IframeExecutionService>) {
+      state.executionService = action.payload;
+    },
+    setPermissionController(
+      state,
+      action: PayloadAction<GenericPermissionController>,
+    ) {
+      state.permissionController = action.payload as any;
+    },
+    setSubjectMetadataController(
+      state,
+      action: PayloadAction<SubjectMetadataController>,
+    ) {
+      state.subjectMetadataController = action.payload as any;
```

Comment:
Does this need a cast?

Reply:
Yes, similar to the `PermissionController`, there's some weirdness with writable drafts.